### PR TITLE
Apply layout for Grover PDF library [SCI-9566]

### PR DESCRIPTION
**Pull request by ivanscinote**
Jira ticket: [SCI-9566](https://scinote.atlassian.net/browse/SCI-9566)

### What was done
- apply styling that works with Grover library
- refactor perform method in pdf_job.rb

[SCI-9566]: https://scinote.atlassian.net/browse/SCI-9566?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/jobs/reports/pdf_job.rb [hunk 1]
```diff
@@ -15,89 +15,173 @@ class PdfJob < ApplicationJob
...
-          end
+      @report = Report.find(report_id)
+      @user = User.find(user_id)
+      @file = Tempfile.new(['report', '.pdf'], binmode: true)
+      initialize_template
+      set_renderer_context
+      generate_pdf_content
+      attach_pdf_to_report
+    rescue StandardError => e
+      handle_error(e)
```
Comment (okriuchykhin) [inline app/jobs/reports/pdf_job.rb:26]: keep handling code here

File: app/jobs/reports/pdf_job.rb [hunk 2]
```diff
@@ -15,89 +15,173 @@ class PdfJob < ApplicationJob
...
+    ensure
+      cleanup
```
Comment (okriuchykhin) [inline app/jobs/reports/pdf_job.rb:28]: and same here

File: app/jobs/reports/pdf_job.rb [hunk 3]
```diff
@@ -15,89 +15,173 @@ class PdfJob < ApplicationJob
...
-          else
-            Extends::REPORT_TEMPLATES.keys.first.to_s
-          end
+      @report = Report.find(report_id)
+      @user = User.find(user_id)
+      @file = Tempfile.new(['report', '.pdf'], binmode: true)
+      initialize_template
+      set_renderer_context
+      generate_pdf_content
+      attach_pdf_to_report
```
Comment (okriuchykhin) [inline app/jobs/reports/pdf_job.rb:24]: it does much more than just attach

File: app/jobs/reports/pdf_job.rb [hunk 4]
```diff
@@ -15,89 +15,173 @@ class PdfJob < ApplicationJob
...
+        'reports/content',
+        'reports/template_helpers'
+      ]
+
+      @style_tag_options = []
+      @script_tag_options = []
+
+      css_files.each do |file_name|
+        css_content = fetch_asset_content("#{file_name}.css")
+        @style_tag_options << { content: css_content }
```
Comment (artoscinote) [inline app/jobs/reports/pdf_job.rb:112]: can be written as

```
@style_tag_options = css_files.map do |file_name|
  { content: fetch_asset_content("#{file_name}.css") }
end
```

File: app/jobs/reports/pdf_job.rb [hunk 5]
```diff
@@ -15,89 +15,173 @@ class PdfJob < ApplicationJob
...
       end
-    rescue StandardError => e
-      raise e if report.blank?
+
+      @style_tag_options.concat(font_awesome_links)
+
+      javascript_files.each do |file_name|
+        script_content = fetch_asset_content("#{file_name}.js")
+        @script_tag_options << { content: script_content }
+      end
```
Comment (artoscinote) [inline app/jobs/reports/pdf_job.rb:120]:
```
@script_tag_options = javascript_files.map do |file_name|
  { content: fetch_asset_content("#{file_name}.js") }
end
```

**Review by artoscinote — CHANGES_REQUESTED** [review]

File: app/jobs/reports/pdf_job.rb [hunk 6]
```diff
@@ -149,10 +233,8 @@ def merge_pdf_files(file, report_file)
       merged_file
     end
 
-    def prepend_title_page(file, template, report, renderer)
-      unless File.exist?(Rails.root.join('app', 'views', 'reports', 'templates', template, 'cover'))
-        return file
-      end
+    def prepend_title_page(file, template, report, renderer, has_cover, style_tag_options, script_tag_options)
+      return file unless has_cover
```
Comment (okriuchykhin) [inline app/jobs/reports/pdf_job.rb:237]: move this condition out

File: app/jobs/reports/pdf_job.rb [hunk 7]
```diff
@@ -15,90 +15,166 @@ class PdfJob < ApplicationJob
...
+        emulate_media: 'screen',
+        display_url: "#{Rails.application.config.force_ssl ? 'https' : 'http'}://" \
+                     "#{Rails.application.secrets.mail_server_url}"
+      ).to_pdf(@file.path)
+    end
+
+    def process_attach_pdf_report_and_notify
+      @file.rewind
+      if @has_cover
+        @file = prepend_title_page(@file, @template, @report, @renderer, @style_tag_options, @script_tag_options)
```
Comment (okriuchykhin) [inline app/jobs/reports/pdf_job.rb:155]: these are all instance variables, no need to pass them as params

File: app/jobs/reports/pdf_job.rb [hunk 8]
```diff
@@ -15,90 +15,166 @@ class PdfJob < ApplicationJob
...
+        print_background: true,
+        margin: current_margin,
+        display_header_footer: true,
+        header_template: @header_html,
+        footer_template: @footer_html,
+        style_tag_options: @style_tag_options,
+        script_tag_options: @script_tag_options,
+        page_ranges: "#{cover_pages_shift}-999999",
+        emulate_media: 'screen',
+        display_url: "#{Rails.application.config.force_ssl ? 'https' : 'http'}://" \
```
Comment (okriuchykhin) [inline app/jobs/reports/pdf_job.rb:147]: Rails.application.routes.default_url_options[:host]

**Review by okriuchykhin — CHANGES_REQUESTED** [review]
check how you use instance variables

File: app/jobs/reports/pdf_job.rb [hunk 9]
```diff
@@ -165,27 +235,33 @@ def prepend_title_page(file, template, report, renderer)
...
       Grover.new(
         title_page_html,
-        format: 'A4'
+        format: 'A4',
+        style_tag_options: @style_tag_options,
+        script_tag_options: @script_tag_options,
+        emulate_media: 'screen',
+        print_background: true,
+        display_url: "#{Rails.application.config.force_ssl ? 'https' : 'http'}://" \
+                     "#{Rails.application.secrets.mail_server_url}"
```
Comment (okriuchykhin) [inline app/jobs/reports/pdf_job.rb:253]: Rails.application.routes.default_url_options[:host]

**Review by artoscinote — APPROVED** [review]

**Review by okriuchykhin — APPROVED** [review]

**Pull request merged**